app/models/puppetclass.rb

```diff
@@ -87,31 +84,25 @@ def klass
     name.gsub(module_name+"::","")
   end
 
-  # return host ids from config groups by type
-  def host_ids_from_config_groups(host_type)
-    ids = config_groups.joins(:host_config_groups)
-                 .where("host_config_groups.host_type='#{host_type}'")
-                 .pluck('host_config_groups.host_id') unless config_group_classes.empty?
-    ids || []
-  end
-
   def all_hostgroups(with_descendants = true)
-    ids = hostgroup_ids
-    ids += host_ids_from_config_groups('Hostgroup')
-    hgs = Hostgroup.unscoped.where(:id => ids.uniq)
+    hgs = Hostgroup.authorized
+                   .reorder('')
+                   .eager_load(:hostgroup_classes, :config_groups => [:config_group_classes])
+                   .where("#{self.id} IN (hostgroup_classes.puppetclass_id, config_group_classes.puppetclass_id)")
+                   .uniq
```

Comment:
Good, few points.

We are breaking the contract, it used to be sorted, now it's not. The method is used once (in a RABL), can you add an optional parameter `unsorted` and use it appropriately to save the extra sorting when doing count?

Have we lost `host_type` from above? It's not part of the new query anymore.